### PR TITLE
Improve compatibility with some ipv4 networks.

### PR DESCRIPTION
**Pull request by lprylli**
- dnsmasq has been seen to null-terminate the bootfile option.
- for the gateway/dns information to be present in final packet,
the Router option should be queried again in request as
in discover (which matches behavior of udhcpc/dhclient).

Tested: pxeboot with u-root on dnsmask/ipv4 client.

### CODE REVIEW

**Comment by insomniacslk** [comment]
Thanks @lprylli ! Can you please amend the commit with DCO?

File: dhcpv4/dhcpv4.go [hunk 1]
```diff
@@ -574,7 +580,8 @@ func (d *DHCPv4) RootPath() string {
 //
 // The Bootfile Name option is described by RFC 2132, Section 9.5.
 func (d *DHCPv4) BootFileNameOption() string {
-	return GetString(OptionBootfileName, d.Options)
+	name := GetString(OptionBootfileName, d.Options)
+	return strings.TrimRight(name, "\x00")
```
Comment (insomniacslk) [inline dhcpv4/dhcpv4.go]: that's sad.. I wonder if we should notify the user of this or silently swallow the null bytes. Thank you null-terminated strings? :(

Comment (lprylli) [reply dhcpv4/dhcpv4.go]: As RFC2132 states "Options containing NVT ASCII data SHOULD NOT include a trailing NULL; however, the receiver of such options MUST be prepared to delete trailing nulls if they exist." :-)

Comment (insomniacslk) [reply dhcpv4/dhcpv4.go]: sounds good :)

**Comment by insomniacslk** [comment]
Alright, the PR looks good. I'll wait for tests and DCO before merging

**Comment by insomniacslk** [comment]
Also, we should look for other options using NVT ASCII and do the same. Maybe a separate PR is ok

File: dhcpv4/dhcpv4.go (hunk 1 above)
Comment (hugelgupf) [inline dhcpv4/dhcpv4.go]: Do the same for the TFTPServerName option?

Comment (lprylli) [reply dhcpv4/dhcpv4.go]: done

**Comment by lprylli** [comment]
Adding tftpservername as suggested (it is really symmetric in that it is the only one dnsmasq also treats specially and null-terminate, although probably not very commonly used).

Added DCO.

There is probably a case for moving up a level and doing null-trimming in GetString, but I was trying to be conservative, minimize the risk of regression, and the scope of changes to case I tested in a "real" common scenario with dnsmasq. So handling all NVT-ASCII string might be better done in separate PR.

**Comment by lprylli** [comment]
Did an extra push (

**Comment by pmazzini** [comment]
Why did you close the PR?

**Comment by lprylli** [comment]
sorry fat finger, not that much practive with github PR, pressed the wrong "comment and close" button instead of just "comment" intended.

**Comment by insomniacslk** [comment]
Thanks @lprylli, I'll review again tomorrow during daytime

**Comment by pmazzini** [comment]
I am not merging it yet to let @insomniacslk have a look as well.

**Pull request merged**